### PR TITLE
feat: add handler for smart contract events

### PR DESCRIPTION
**Pull request by djordon**
## Description

Closes https://github.com/stacks-network/sbtc/issues/426. This wraps up the handler for stacks print events.

## Changes

* Add the network to the list of options in the signer's config file.
* Have the private key in the config file to be the generic signer private key.
* Add new tables for our print events from the `sbtc-registry` smart contract.
* Add functions for writing the above print events to the database.
* Add test fixtures for the various print events and unit tests that use them.

## Testing Information

This includes a unit test that checks the deserialization of the webhooks emitted from stacks core. This payload was sent from a stacks core node under Nakamoto on commit https://github.com/stacks-network/stacks-core/commit/cbf0c52fb7a0b8ac55badadcf3773ca0848a25cf which was made on 2024-08-20. The raw body of the webhook was stored as the test fixtures added in this PR. These events were triggered by running test cases in the `complete_deposit_wrapper_tx_accepted` integration test. The stacks node was configured to emit events like so:

```toml
[[events_observer]]
endpoint = "host.docker.internal:8801"
retry_count = 3
include_data_events = false
events_keys = [
    "SN2V7WTJ7BHR03MPHZ1C9A9ZR6NZGR4WM8HT4V67Y.sbtc-registry::print",
]
```

This testing setup followed the same testing setup as in https://github.com/stacks-network/sbtc/pull/303 and https://github.com/stacks-network/sbtc/pull/442. 

## Checklist:

- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: signer/src/api/new_block.rs
Comment (cylewitruk) [inline signer/src/api/new_block.rs:44]: I think this is fine for now since realistically this shouldn't happen (unless Stacks makes a release that completely borks the json schema... totally possible :sweat_smile: ), but in reality this is probably a _fatal_ situation, isn't it? If we're missing blocks then we're potentially missing important sbtc contract calls?

File: signer/src/api/new_block.rs
Comment (cylewitruk) [inline signer/src/api/new_block.rs:122]: Maybe we should actually make this a const in `config` (i.e. move the const from `config::tests` that i added to the `config` module with a `#[cfg(test)]` on it), since it seems we'll be using this in multiple places.

File: signer/src/config.rs
Comment (cylewitruk) [inline signer/src/config.rs:191]: :+1:

File: signer/src/config/default.toml
Comment (cylewitruk) [inline signer/src/config/default.toml:57]: I'd add:
```
# Possible Values: mainnet, testnet, regtest
# Environment: SIGNER_SIGNER__NETWORK
```

File: signer/tests/integration/contracts.rs
Comment (cylewitruk) [inline signer/tests/integration/contracts.rs:131]: Tbh it wouldn't hurt to have this in the actual `StacksClient`, you're already half-using it

File: signer/tests/integration/postgres.rs
Comment (cylewitruk) [inline signer/tests/integration/postgres.rs:718]: Not something for this PR, but I would suggest moving from `fake` to [proptest](https://crates.io/crates/proptest) [strategies](https://altsysrq.github.io/proptest-book/proptest/tutorial/strategy-basics.html) and writing [proptests](https://altsysrq.github.io/proptest-book/proptest/tutorial/macro-proptest.html) instead. It provides much more robust testing using a multitude of different randomized values w/value shrinking etc. vs. the one random value you get here.

**Review by cylewitruk — COMMENTED** [review]
Lgtm, just had a few small comments/suggestions

File: signer/src/api/new_block.rs
Comment (djordon) [reply signer/src/api/new_block.rs:122]: Good idea, will do.

File: signer/src/api/new_block.rs
Comment (djordon) [reply signer/src/api/new_block.rs:44]: Yeah that makes sense, this shouldn't happen. Although it's more likely that we have some "bug" in our expectation of the schema. If we hit this we'd have to have the stacks node replay the events somehow (no idea how to actually do that).

File: signer/tests/integration/contracts.rs
Comment (djordon) [reply signer/tests/integration/contracts.rs:131]: Honestly, I got lazy and didn't want to add the tests that I normally would for a proper function for the `StacksClient` 😅. Plus I wanted to keep this PR focused. But this is a good point, I'll add it in another PR though.

Comment (cylewitruk) [reply signer/tests/integration/contracts.rs:131]: Haha okay, was just a suggestion anyway, i'll resolve this comment then :)

File: signer/src/api/new_block.rs
Comment (cylewitruk) [reply signer/src/api/new_block.rs:44]: Something in me feels like there maybe should be some p2p chatter around sbtc events, so a signer (and its peers) can determine if it's lagging/faulty... which should probably result in the signer being ejected from the active set?

File: signer/tests/integration/postgres.rs
Comment (djordon) [reply signer/tests/integration/postgres.rs:718]: Yeah, that's a good idea. I'll give it look.

File: signer/src/api/new_block.rs
Comment (djordon) [reply signer/src/api/new_block.rs]: I ended up moving it to the testing module, in case we need it for our integration tests.

File: signer/src/config/default.toml
Comment (djordon) [reply signer/src/config/default.toml:57]: Updated in https://github.com/stacks-network/sbtc/pull/474/commits/57efea14a608dde5e06ab2393a77221ae3b2dde4

File: signer/src/api/new_block.rs
Comment (djordon) [reply signer/src/api/new_block.rs:44]: > This isn't possible today :( The events are only emitted as a part of block processing. 

That makes sense. Is there a way to have a node forget/delete blocks past a certain height (this is closer to what I had in mind)? If not we'd have to wipe it clean? Can we do that in a "nice" way?

> Something in me feels like there maybe should be some p2p chatter around sbtc events, so a signer (and its peers) can determine if it's lagging/faulty... which should probably result in the signer being ejected from the active set?

The chatter idea might be a good one 🤔. But they probably shouldn't be ejected if they are falling behind, we've just been designing the signers with that possibility in mind.

Comment (cylewitruk) [reply signer/src/api/new_block.rs:44]: > That makes sense. Is there a way to have a node forget/delete blocks after a certain height (this is closer to what I had in mind)? If not we'd have to wipe it clean? Can we do that in a "nice" way?

Unfortunately, no :frowning_face:  The concept of "deleting" data is a bit anti-blockchain, so I can understand why rolling-back a node hasn't been part of the design considerations. I'm sure it'd be possible to implement, but I'm not so sure stacks-core would go along with it.

>The chatter idea might be a good one 🤔. But they probably shouldn't be ejected if they are falling behind, we've just been designing the signers with that possibility in mind.

Are there some design docs I could dig into regarding this design so I can get a better idea how it's handled?

Comment (cylewitruk) [reply signer/src/api/new_block.rs:44]: I'll go ahead and give an approve so you don't get blocked on this discussion, but maybe we should create an issue to track this?

**Review by cylewitruk — APPROVED** [review]
Approving so you don't get blocked, but we should probably create an issue to track the "what happens if event-observer deserialization fails" -- if nothing else than that we can land in a documented routine/workaround in the unlikely event of..

File: signer/src/api/new_block.rs
Comment (netrome) [inline signer/src/api/new_block.rs:122]: Nit: `func` is a quite uninformative name. Perhaps we could rename it to `table_is_empty`?

File: signer/src/testing/dummy.rs
Comment (netrome) [inline signer/src/testing/dummy.rs:260]: I don't think we'd want to do a blanket implementation here. This prevents any other implementations for other fakers. Instead, I'd recommend specifying this to just the `fake::Faker` type:

```suggestion
impl fake::Dummy<fake::Faker> for WithdrawalAcceptEvent {
```

File: signer/src/testing/dummy.rs
Comment (netrome) [inline signer/src/testing/dummy.rs:276]:
```suggestion
impl fake::Dummy<fake::Faker> for WithdrawalRejectEvent {
```

File: signer/src/testing/dummy.rs
Comment (netrome) [inline signer/src/testing/dummy.rs:287]:
```suggestion
impl fake::Dummy<fake::Faker> for WithdrawalCreateEvent {
```

File: signer/src/testing/dummy.rs
Comment (netrome) [inline signer/src/testing/dummy.rs:306]:
```suggestion
impl fake::Dummy<fake::Faker> for CompletedDepositEvent {
```

**Review by netrome — APPROVED** [review]
Looks good to me, but I think the Fake implementations can be specific rather than using blanket implementations.

File: signer/src/api/new_block.rs
Comment (djordon) [reply signer/src/api/new_block.rs:122]: Great idea. Will update.

File: signer/src/api/new_block.rs
Comment (djordon) [reply signer/src/api/new_block.rs:44]: > I'll go ahead and give an approve so you don't get blocked on this discussion, but maybe we should create an issue to track this?

Yeah, good point, I'll open one for adding metrics and alerts. That's how I would handle this case.

Comment (djordon) [reply signer/src/api/new_block.rs:44]: I opened https://github.com/stacks-network/sbtc/issues/481, but it needs some refining.

File: signer/src/testing/dummy.rs
Comment (djordon) [reply signer/src/testing/dummy.rs:260]: Oh right 🤦🏽 good point.

**Comment by djordon** [comment]
> Approving so you don't get blocked, but we should probably create an issue to track the "what happens if event-observer deserialization fails" -- if for nothing else than that we can land in a documented routine/workaround in the unlikely event of..

Good point, I added https://github.com/stacks-network/sbtc/issues/481 for this. Also, we'll need to break down that ticket further.

File: signer/src/api/new_block.rs
Comment (djordon) [reply signer/src/api/new_block.rs:44]: > I'll go ahead and give an approve so you don't get blocked on this discussion, but maybe we should create an issue to track this?

There are a few. For this particular question I think the designs are laid out in https://github.com/stacks-network/sbtc/issues/39 and https://github.com/stacks-network/sbtc/issues/44. There are proper google docs too, with their links in different issues. I generally find them using this query https://github.com/stacks-network/sbtc/issues?q=is%3Aissue+design+is%3Aclosed+LLD, their usually the tickets with LLD in the name.

**Pull request merged**
